chain/checkpointing/sub.go

```diff
@@ -0,0 +1,506 @@
+			if err != nil {
+				return nil, err
+			}
+			err = p.UnmarshalBinary(pByte)
+			if err != nil {
+				return nil, err
+			}
+			verificationShares[party.ID(key)] = &p
+		}
+
+		config = &keygen.TaprootConfig{
+			ID:                 party.ID(host.ID().String()),
+			Threshold:          configTOML.Thershold,
+			PrivateShare:       &privateShare,
+			PublicKey:          publickey,
+			VerificationShares: verificationShares,
+		}
+	}
+
+	return &CheckpointingSub{
+		pubsub:    pubsub,
+		topic:     nil,
+		sub:       nil,
+		host:      host,
+		api:       &api,
+		events:    e,
+		init:      false,
+		ptxid:     "",
+		config:    config,
+		newconfig: nil,
+	}, nil
+}
+
+func (c *CheckpointingSub) listenCheckpointEvents(ctx context.Context) {
+
+	checkFunc := func(ctx context.Context, ts *types.TipSet) (done bool, more bool, err error) {
+		return false, true, nil
+	}
+
+	changeHandler := func(oldTs, newTs *types.TipSet, states events.StateChange, curH abi.ChainEpoch) (more bool, err error) {
+		log.Infow("State change detected for power actor")
+
+		idsStrings := c.orderParticipantsList()
+
+		fmt.Println("Participants list :", idsStrings)
+
+		ids := c.formIDSlice(idsStrings)
+
+		id := party.ID(c.host.ID().String())
+
+		threshold := 2
+		n := NewNetwork(ids, c.sub, c.topic)
+		f := frost.KeygenTaproot(id, ids, threshold)
+
+		handler, err := protocol.NewMultiHandler(f, []byte{1, 2, 3})
+		if err != nil {
+			fmt.Println(err)
+			panic(err)
+		}
+		c.LoopHandler(ctx, handler, n)
+		r, err := handler.Result()
+		if err != nil {
+			fmt.Println(err)
+			panic(err)
+		}
+		fmt.Println("Result :", r)
+
+		c.newconfig = r.(*keygen.TaprootConfig)
+
+		return true, nil
+	}
+
+	revertHandler := func(ctx context.Context, ts *types.TipSet) error {
+		return nil
+	}
+
+	match := func(oldTs, newTs *types.TipSet) (bool, events.StateChange, error) {
+		newAct, err := c.api.StateGetActor(ctx, mpower.PowerActorAddr, newTs.Key())
+		if err != nil {
+			return false, nil, err
+		}
+
+		oldAct, err := c.api.StateGetActor(ctx, mpower.PowerActorAddr, oldTs.Key())
+		if err != nil {
+			return false, nil, err
+		}
+
+		var oldSt, newSt mpower.State
+
+		bs := blockstore.NewAPIBlockstore(c.api)
+		cst := cbor.NewCborStore(bs)
+		if err := cst.Get(ctx, oldAct.Head, &oldSt); err != nil {
+			return false, nil, err
+		}
+		if err := cst.Get(ctx, newAct.Head, &newSt); err != nil {
+			return false, nil, err
+		}
+
+		// If Power Actors list has changed start DKG
+		if !c.init {
+			ts, err := c.api.ChainGetTipSetByHeight(ctx, 0, oldTs.Key())
+			if err != nil {
+				panic(err)
+			}
+			data := ts.Cids()[0]
+			err = c.initiate(data.Bytes())
+			if err != nil {
+				panic(err)
+			}
+			return false, nil, nil
+		}
+
+		// ZONDAX TODO
+		// Activate checkpointing every 20 blocks
+		fmt.Println("Height:", newTs.Height())
+		if newTs.Height()%20 == 0 {
+			fmt.Println("Check point time")
+
+			// Initiation and config should be happening at start
```

Comment:
We may need to revise this logic to be sure that:
* The checkpoint is running every 20 blocks.
* We may need to wait a few epochs before detecting a checkpointing epoch to ensure finality (the tipset we detect in the event change may be reverted and not be final, and committing a reverted tipset in a checkpoint may be quite bad).
* We should check that we are only committing the checkpoints one. Due to the way `checkFunc` works, it may trigger the `ChangeFunc` every time the state change is detected, which could end up translating in a checkpoint being committed several times.

Just wanted to note it here so we test it at some point.

Reply:
Also, we should probably add a config parameter to configure the checkpointing period.

Reply:
Re-reading the function again, it may be a good idea to detect the change here, wait for a few epochs, and the trigger the changeFunc to create a new checkpoint. That way we ensure finality. I do something similar here in case you need inspiration: https://github.com/filecoin-project/eudico/blob/d08ea35d0e4bfe437b7d1910829300fbb49a18f0/chain/consensus/hierarchical/subnet/events.go#L203